pkg/dispatcher/dispatcher.go

```diff
@@ -100,16 +103,45 @@ func (p *Dispatcher) dispatch(event events.SchedulingEvent) error {
 	if !p.isRunning() {
 		return fmt.Errorf("dispatcher is not running")
 	}
-	for p.isRunning() {
-		select {
-		case p.eventChan <- event:
-			return nil
-		case <-time.After(time.Second * 3):
-			log.Logger.Error("event channel is full, keep waiting",
-				zap.Int("channelLength", len(p.eventChan)))
-		}
+	select {
+	case p.eventChan <- event:
+		log.Logger.Info("dispatched", zap.Int("capacity", cap(p.eventChan)),
+			zap.Int("length", len(p.eventChan)))
+		return nil
+	default:
+		log.Logger.Warn("event channel is full, transition to async-dispatch mode",
+			zap.Int("channelLength", len(p.eventChan)),
+			zap.String("event", fmt.Sprintf("%+v", event)))
+		p.asyncDispatch(event)
```

Comment:
I think we need to set a limit here, to avoid this creating too many goroutines. Can we create a limit counter here, lets set a default value of 10,000. If num of goroutines goes beyond this value, then panic. This is just to protect some worst-case scenarios.

Reply:
Updates: 
(1) Set a limit for async-dispatch goroutines in Dispatcher#asyncDispatch.
(2) Remove channelLength and event fields from this log, add asyncDispatchCount field and move this log into Dispatcher#asyncDispatch.
(3) Add log in Dispatcher#init to print EventChannelCapacity and DispatchTimeoutInSeconds.